### PR TITLE
test: CalcJoinPoolShares unit tests with single token by sharing them with calcSingleAssetJoin

### PR DESCRIPTION
**Pull request by xBalbinus**
## What is the purpose of the change

`CalcJoinPoolShares` single token flow calls into `calcSingleAssetJoin`. As a result, we can share test cases to unit test both at the same time.

This PR refactors `TestCalcJoinPoolShares` and `TestCalcSingleAssetJoin` to share test cases. Only 2 new test cases are added in this PR (marked with comments). The rest are copied over.

## Testing and Verifying
- `go test -timeout 30s -run ^TestCalcSingleAssetJoin$ github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/balancer`
- `go test -timeout 30s -run ^TestCalcJoinPoolShares$ github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/balancer`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Review by alexanderbez — COMMENTED** [review]
It seems this has overlap with @p0mvn's PR...?

**Comment by xBalbinus** [comment]
Some, but not all. I just tried to sync mine with Roman’s

**Comment by p0mvn** [comment]
I will add on to this PR shortly

File: x/gamm/pool-models/balancer/amm_test.go [hunk 1]
```diff
@@ -14,12 +14,275 @@ import (
...
+		// where:
+		// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+		//	A_t = amount of deposited asset = 50,000
+		//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+		//	W_t = normalized weight of deposited asset in pool = 0.5 (equal weights)
+		// 	swapFeeRatio = (1 - (1 - W_t) * swapFee)
+		// Plugging all of this in, we get:
+		// 	Full solution: https://www.wolframalpha.com/input?i=%28100+*+10%5E18+%29*+%28%28+1+%2B+%2850%2C000+*+%281+-+%281+-+0.5%29+*+0.99%29+%2F+1000000000000%29%29%5E0.5+-+1%29
+		// 	Simplified:  P_issued = 1_262_500_000_000
+		name:    "single tokensIn - equal weights with 0.99 swap fee",
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm_test.go:117]: new test case

File: x/gamm/pool-models/balancer/amm_test.go [hunk 2]
```diff
@@ -14,12 +14,275 @@ import (
...
+		// where:
+		// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+		//	A_t = amount of deposited asset = 50,000
+		//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+		//	W_t = normalized weight of deposited asset in pool = 0.25 (asset A, uosmo, has weight 1/4 of uatom)
+		// 	swapFeeRatio = (1 - (1 - W_t) * swapFee)
+		// Plugging all of this in, we get:
+		// 	Full solution: https://www.wolframalpha.com/input?i=%28100+*+10%5E18+%29*+%28%28+1+%2B+%2850%2C000+*+%281+-+%281+-+0.25%29+*+0.99%29+%2F+1000000000000%29%29%5E0.25+-+1%29
+		// 	Simplified:  P_issued = 321_875_000_000
+		name:    "single tokensIn - unequal weights with 0.99 swap fee",
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm_test.go:148]: new test case

File: x/gamm/pool-models/balancer/amm_test.go [hunk 3]
```diff
@@ -14,12 +14,275 @@ import (
...
+}
+
 // allowedErrRatio is the maximal multiplicative difference in either
 // direction (positive or negative) that we accept to tolerate in
 // unit tests for calcuating the number of shares to be returned by
 // joining a pool. The comparison is done between Wolfram estimates and our AMM logic.
 const allowedErrRatio = "0.0000001"
 
+// see calcJoinSharesTestCase struct definition.
+var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm_test.go:40]: Commenting to create an issue for further work.

More test cases should be added in a separate PR:
- with error return
- with another value of `tokensIn` (we consistently use 50000 which isn't the best practice

**Review by alexanderbez — APPROVED** [review]
